editors/code/src/commands/on_enter.ts

```diff
@@ -19,6 +19,12 @@ async function handleKeypress(ctx: Ctx) {
     const change = await client.sendRequest<undefined | SourceChange>(
         'rust-analyzer/onEnter',
         request,
+    ).catch(
+        (_error: any) => {
+            // FIXME: switch to the more modern (?) typed request infrastructure
```

Comment:
@matklad, yes, this is possible, we should abstract `client.sendRequest()` to 
```ts
function trySendRequest<T extends OurAPITypeInfo>(
    param: T["method"][["param"]
): Promise<T["return-type"]>
```
and encode the type information about our custom LSP api in `OurAPITypeInfo` interface.
I planned to do that refactor after tackling enhanced typing issues...

Reply:
Or use separate functions:
```
function requestInlayHintsWithRetry(client, param: InlayHintsOpts): Promise<InlayHint[]>;
function requestOnEnter(param: OnEnterOpts): Promise<OnEnterResponse>;
```

Reply:
@matklad which do you like most?

Reply:
Taking the first approach we may abstract the type info in an npm package and share it with other ts plugins ( coc.nvim for example), though this may be too much overhead since our custom API is not so big

Reply:
We don't need `OurAPITypeInfo`, we should use existing RPCMethodType: https://github.com/microsoft/vscode-languageserver-node/blob/f425af9de46a0187adb78ec8a46b9b2ce80c5412/client/src/client.ts#L99

Reply:
@matklad , yeah but it doesn't contain types for our custom methods like `inlayHints` and `onEnter` , right? Might I ask you to show an example how you would make it this way?